### PR TITLE
TPC seeding overhaul

### PR DESCRIPTION
**Pull request by petermic**
[comment]: <> (Please tell us something about this pull request)

## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [x] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## What kind of change does this PR introduce? (Bug fix, feature, ...)https://github.com/sPHENIX-Collaboration/macros/pull/556

[comment]: <> ( What does this PR do? Linking to talk in software meeting encouraged )
Rewrites PHCASeeding, ALICEKF, and PHSimpleKFProp to be more efficient, especially at lower pT and in pp data.
- Switches from eta-phi search window to z-phi search window
- Builds multiple cluster chains if multiple compatible clusters are found
- Propagates and filters using energy loss corrections
- Propagates with double previous resolution

Note that the change to a z-phi window involves a significant change to the G4_Tracking.C settings for PHCASeeding: https://github.com/sPHENIX-Collaboration/macros/pull/556.

## TODOs (if applicable)

[comment]: <> ( In case this is a draft PR, e.g. for running checks using Jenkins, please make the pull request as a draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/  )

## Links to other PRs in macros and calibration repositories (if applicable)

https://github.com/sPHENIX-Collaboration/macros/pull/556.

### CODE REVIEW

**Comment by petermic** [comment]
The gcc-new build is marked as "failed", but Jenkins tells me no new issues, and the build log shows a successful build, as far as I can tell.

**Comment by osbornjd** [comment]
It looks like the vertexing QA jupyter notebook dies, however the other notebooks run seemingly without an issue:

https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/test-tracking-low-occupancy-qa/2552/consoleFull

@blackcathj any ideas?

**Comment by blackcathj** [comment]
> It looks like the vertexing QA jupyter notebook dies, however the other notebooks run seemingly without an issue:
> 
> https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/test-tracking-low-occupancy-qa/2552/consoleFull
> 
> @blackcathj any ideas?

The processing of the notebook timed out, suggesting the underlying ROOT process get stuck. ROOT is not very graceful when bump into problem, in particular with pointers. Unfortunately, it would require me to reproduce the QA root files and run the notebook interactively to find out why

**Comment by petermic** [comment]
> It looks like the vertexing QA jupyter notebook dies, however the other notebooks run seemingly without an issue:
> 
> https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/test-tracking-low-occupancy-qa/2552/consoleFull
> 
> @blackcathj any ideas?

If this is unrelated to the gcc-new build fail, I wonder if this might be because this is running the old macro with new coresoftware. The previous eta window size, if used as a delta-z window size, is not going to return very many tracks at all (basically only those that have very low pz).

**Comment by petermic** [comment]
Anyway, the gcc-new build is now reporting a new issue on a file I didn't change: https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/Build-Master-gcc8/2923/gcc/new/.

**Comment by blackcathj** [comment]
Yes. That is a merge problem in the `master` branch that we are fixing. We will restart the check manually once the `master` branch is fixed

**Comment by blackcathj** [comment]
Reasons why the QA jobs get stuck appears to be the vertex histograms are mostly empty which fails fit. Then obtaining fit `TF1` returned a null pointer, with which without check will send root kernel into a stuck state.

QA notebooks are updated to handle this scenario with just showing the empty histogram: 

image

**Comment by petermic** [comment]
This makes sense, considering that this QA is being run before merging the changes in macros -- the previous default windows (for delta-eta) are very small in delta-z, so there's likely not enough tracks reconstructed to form a vertex.

**Comment by petermic** [comment]
Anyway, there have been a few commits added to this PR by accident (everything after "Fixed comma deletion", in fact); I managed to revert one of them, but the latest merge appears to introduce new changes. Should I close this PR and start over? Is there a way to fix this? Or do the additional changes from the latest commit not matter?

**Comment by osbornjd** [comment]
You can either rebase your branch and force push the changes to overwrite the history of the last couple of commits or you can open a new PR, whichever you prefer. I'm not sure why the QA didn't run with your macro PR as I set the build to do that, but that is a problem on the Jenkins end so let me try to figure out why that didn't propagate.

**Comment by blackcathj** [comment]
> This makes sense, considering that this QA is being run before merging the changes in macros -- the previous default windows (for delta-eta) are very small in delta-z, so there's likely not enough tracks reconstructed to form a vertex.

Make sense. When you are ready, ping Joe or I and we can manually run CI with your macro PR instead of the default macros

**Comment by petermic** [comment]
Broken beyond repair, starting over.

**Comment by osbornjd** [comment]
Jenkins didn't give us a notification because it was a special build, but for artifacts here is the build [info](
https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/Build-Master-gcc8/2965/) for this PR with the corresponding macros PR.

The low occupancy QA is [here](https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/test-tracking-low-occupancy-qa/2588/)

The tracking QA looks significantly improved in terms of efficiency and number of tracks with >40 TPC [clusters](https://nbviewer.org/github/sPHENIX-Collaboration/QA-gallery/blob/jenkins-sPHENIX-test-tracking-low-occupancy-qa-2588-test-tracking_Event1000_Sum16/QA-tracking.ipynb). The purity is worse at the lowest pT so we may need additional tuning of the TPC ghost seed module, which can be done later (but in fact is improved at ~500 MeV).

Note the degraded momentum resolution is an ongoing issue with the new cluster v4.

@petermic once I get your approval that this does not require any more changes I will merge this with the macros PR change. We can propagate this to the streaming pp mode in a separate PR.

File: offline/packages/trackreco/PHSimpleKFProp.cc [hunk 1]
```diff
@@ -258,7 +257,40 @@ int PHSimpleKFProp::process_event(PHCompositeNode* topNode)
...
+        {
+          dumvec2.push_back(*iter);
+          auto pos = globalPositions.at(*iter);
+          pretrackClusPositions.insert(std::make_pair(*iter,pos));
+        }
+
+      pretrack.circleFitByTaubin(pretrackClusPositions, 7, 55);
+      pretrack.lineFit(pretrackClusPositions, 7, 55);
+
+      new_chains.push_back(PropagateTrack(&pretrack, prepair.second.at(0),
```
Comment (osbornjd) [inline offline/packages/trackreco/PHSimpleKFProp.cc:293]: I can't remember now, but did you check the single track timing after this change? I imagine running the propagation twice adds additional time - it gives us much better performance so for now so be it, but I was also wondering if this is something we should look out for in the HIJING QA.

**Comment by petermic** [comment]
> Jenkins didn't give us a notification because it was a special build, but for artifacts here is the build [info](https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/Build-Master-gcc8/2965/) for this PR with the corresponding macros PR.
> 
> The low occupancy QA is [here](https://web.sdcc.bnl.gov/jenkins-sphenix/job/sPHENIX/job/test-tracking-low-occupancy-qa/2588/)
> 
> The tracking QA looks significantly improved in terms of efficiency and number of tracks with >40 TPC [clusters](https://nbviewer.org/github/sPHENIX-Collaboration/QA-gallery/blob/jenkins-sPHENIX-test-tracking-low-occupancy-qa-2588-test-tracking_Event1000_Sum16/QA-tracking.ipynb). The purity is worse at the lowest pT so we may need additional tuning of the TPC ghost seed module, which can be done later (but in fact is improved at ~500 MeV).
> 
> Note the degraded momentum resolution is an ongoing issue with the new cluster v4.
> 
> @petermic once I get your approval that this does not require any more changes I will merge this with the macros PR change. We can propagate this to the streaming pp mode in a separate PR.

QA looks as I expected it to look, so everything seems ok from my end for the merge.

**Pull request merged**